### PR TITLE
Enhances the DiscountCode API permissions and permission_manager

### LINKED ISSUE #3802: Ensure basic endpoints are accessible without authentication
### Event-related endpoints

- [x] Getting a list of events & details about a single event
- [x] Getting a list of sessions of an event and the details about a single session
- [x] Getting a list of microlocations of an event and the details about a single microlocation
- [x] Getting a list of tracks of an event and the details about a single track
- ~Getting a list of speakers of an event and the details about a single speaker~ [ Edited by Piyush: This need to be done separately since speakers API isn't ready yet ]
- [x] Getting a list of tickets of an event and the details about a single tickets
- [x] Getting call for speakers info of an event
- [x] Getting copyright info of an event

### Other endpoints
- [x] Getting settings (all fields except all client IDs, access tokens, secrets, passwords, email settings). But note, `stripe-publishable-key` is required.
- [x] Getting pages list & detail

### PR DESCRIPTION
**Pull request by poush**
Fixes #3802 and #3990

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `nextgen` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
~Added one fix where NorFoundError used in PR #4013 was a Response Class while we had to use Exception Class. So added it.~
Added Callable helper method `has_access` in permission_manager so that permissions can be used easily in any methods.
Little refactoring.

### CODE REVIEW

File: app/api/discount_codes.py [hunk 1]
```diff
@@ -261,47 +243,39 @@ def before_get_object(self, view_kwargs):
...
             used_for = discount.used_for
 
-        if discount.used_for == 'ticket' and (current_user.is_coorganizer(view_kwargs.get('event_id'))
-           or current_user.is_organizer(view_kwargs.get('event_id') or current_user.is_admin)) and used_for != 'event':
+        if discount.used_for == 'ticket' and has_access('is_coorganizer', event_id=view_kwargs.get('event_id'))\
+           and used_for != 'event':
             self.schema = DiscountCodeSchemaTicket
 
-        elif discount.used_for == 'event' and current_user.is_admin and used_for != 'ticket':
+        elif discount.used_for == 'event' and has_access('is_admin') and used_for != "ticket":
```
Comment (SaptakS) [inline app/api/discount_codes.py]: Just curious. Why replace single quote with double quote in used_for check? 😅

File: app/api/helpers/exceptions.py [hunk 2]
```diff
@@ -9,3 +9,11 @@ class UnprocessableEntity(JsonApiException):
 class ConflictException(JsonApiException):
     title = "Conflict"
     status = 409
+
+
+class NotFoundError(JsonApiException):
```
Comment (SaptakS) [inline app/api/helpers/exceptions.py]: Why add this? We already have ObjectNotFound for this, right?

File: app/api/helpers/permission_manager.py [hunk 3]
```diff
@@ -225,3 +232,12 @@ def permission_manager(view, view_args, view_kwargs, *args, **kwargs):
         return permissions[args[0]](view, view_args, view_kwargs, *args, **kwargs)
     else:
         return ForbiddenError({'source': ''}, 'Access forbidden').respond()
+
+
+def has_access(access_level, **kwargs):
+    if access_level in permissions:
+        auth = permissions[access_level](lambda *a, **b: 100 is 100, (), {}, (), **kwargs)
```
Comment (SaptakS) [inline app/api/helpers/permission_manager.py]: why do you have a check 100 is 100?

File: app/api/discount_codes.py (hunk 1 above)
Comment (poush) [inline app/api/discount_codes.py]: Actually not changed 😅 maybe I wrote it again so added single quotes. should I change them to double?

File: app/api/helpers/exceptions.py (hunk 2 above)
Comment (poush) [inline app/api/helpers/exceptions.py]: I thought I removed it after Medo's PR. I will just update it Thanks! :)

File: app/api/helpers/permission_manager.py (hunk 3 above)
Comment (poush) [inline app/api/helpers/permission_manager.py]: Actually lambda function has to return for any input here so I created an expression to return true but yeah I just tried writing "True" as expression and it worked. Please check.

**Comment by poush** [comment]
@SaptakS Done :)

File: app/api/helpers/permission_manager.py [hunk 4]
```diff
@@ -225,3 +232,12 @@ def permission_manager(view, view_args, view_kwargs, *args, **kwargs):
...
     else:
         return ForbiddenError({'source': ''}, 'Access forbidden').respond()
+
+
+def has_access(access_level, **kwargs):
+    if access_level in permissions:
+        auth = permissions[access_level](lambda *a, **b: True, (), {}, (), **kwargs)
+        if type(auth) is bool and auth is True:
+            return True
+        return False
```
Comment (niranjan94) [inline app/api/helpers/permission_manager.py]: can be removed

Comment (poush) [inline app/api/helpers/permission_manager.py]: Done.

**Comment by poush** [comment]
@niranjan94 @SaptakS @magdalenesuo Reviews Please.

**Pull request merged**